app/models/decision.rb

```diff
@@ -1,50 +1,23 @@
 class Decision
   include ActiveModel::Model
 
-  attr_accessor :variant, :split, :admin
+  attr_accessor :split, :variant
 
-  validates :variant, presence: true, allow_blank: false
-  validates :split, presence: true
-  validates :admin, presence: true
-
-  validate :variant_belongs_to_split
+  validates :split, :variant, presence: true
 
   def save!
-    raise errors.full_messages.to_sentence unless valid?
-
-    weight_split_fully_to_variant
-    reassign_existing_assignments
-
-    true
-  end
-
-  def count
-    raise "count unavailable for unsaved Decision" unless bulk_assignment.persisted?
-    bulk_assignment.assignments.count
-  end
-
-  private
-
-  def weight_split_fully_to_variant
-    split.reweight!(variant => 100)
-  end
-
-  def reassign_existing_assignments
-    bulk_assignment.save!
-    existing_assignments.find_in_batches do |subset|
-      BulkReassignment.create!(assignments: subset, bulk_assignment: bulk_assignment)
+    raise "Variant must be present in the split" unless split.has_variant?(variant)
+    build_split_creation.save!.tap do
+      split.reload
```

Comment:
Feels like knowing to reload could be `SplitCreation`'s responsibility in the upsert case.

Reply:
but then it'd need to know it was an update of an instance, which it doesn't. Unless I add a `#create_split_creation!` method to `Split`, which maybe isn't worth it?

Reply:
It could definitely look at `split.new_record?` before calling save on it. Whether it _should_ gets a little hand-wavey, but typically I see an explicit `reload` as a state-related smell and I'd want to push it as close to the persistence operation as possible (if not fix whatever bug isn't properly setting state during persistence).

Reply:
The problem is that the instance in the upsert is always new, right? I don't get to reload the split instance that the caller has reference to from the split_creation no matter what, unless some identity map stuff has been completed that I'm unaware of in modern rails?

Reply:
really what's happening here is we're favoring reuse of simple upsert codepath over enough decoupling to make reloading unnecessary. I think I might rip out `Split#build_split_creation` in favor of a `Split.reconfigure!` method that secretly encapsulates `SplitCreation` and does the reloading nonsense. I can do that in this PR and see how it looks.

Reply:
Oh, I see. Yes, that is correct, it's not gonna mutate `split` because the way `split.build_a_thing_that_creates_me_even_though_i_exist` (do you like my naming suggestion :trollface:) were smart enough to short-circuit the "find_or_initialize" and just hand it `self`. Which doesn't do.

So you're totally right, this needs an explicit reload 👍